File: src/services/locationService.js
import * as Location from 'expo-location';
import { AppConstants } from '../utils/constants';

export const getCurrentPosition = async () => {
  const { status } = await Location.requestForegroundPermissionsAsync();
  
  if (status !== 'granted') {
    throw new Error('위치 권한이 거부되었습니다.');
  }

  return await Location.getCurrentPositionAsync({
    accuracy: Location.Accuracy.High,
  });
};

export const isWithinCampus = async () => {
  try {
    const position = await getCurrentPosition();
    const distance = calculateDistance(
      AppConstants.campusLatitude,
      AppConstants.campusLongitude,
      position.coords.latitude,
      position.coords.longitude
    );

    return distance <= AppConstants.campusRadiusMeters;
  } catch (error) {
    console.error('위치 확인 오류:', error);
    return false;
  }
};

const calculateDistance = (lat1, lon1, lat2, lon2) => {
  const R = 6371e3; // 지구 반경 (미터)
  const φ1 = lat1 * Math.PI / 180;
  const φ2 = lat2 * Math.PI / 180;
  const Δφ = (lat2 - lat1) * Math.PI / 180;
  const Δλ = (lon2 - lon1) * Math.PI / 180;

  const a = Math.sin(Δφ/2) * Math.sin(Δφ/2) +
          Math.cos(φ1) * Math.cos(φ2) *
          Math.sin(Δλ/2) * Math.sin(Δλ/2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));

  return R * c; // 미터 단위
};

File: lib/services/location_service.dart
import 'package:geolocator/geolocator.dart';
import '../utils/constants.dart';

class LocationService {
  // 현재 위치 가져오기
  Future<Position> getCurrentPosition() async {
    bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
    if (!serviceEnabled) {
      throw Exception('위치 서비스가 비활성화되어 있습니다.');
    }

    LocationPermission permission = await Geolocator.checkPermission();
    if (permission == LocationPermission.denied) {
      permission = await Geolocator.requestPermission();
      if (permission == LocationPermission.denied) {
        throw Exception('위치 권한이 거부되었습니다.');
      }
    }

    if (permission == LocationPermission.deniedForever) {
      throw Exception('위치 권한이 영구적으로 거부되었습니다.');
    }

    return await Geolocator.getCurrentPosition(
      desiredAccuracy: LocationAccuracy.high,
    );
  }

  // 캠퍼스 범위 내에 있는지 확인
  Future<bool> isWithinCampus() async {
    try {
      final position = await getCurrentPosition();
      final distance = Geolocator.distanceBetween(
        AppConstants.campusLatitude,
        AppConstants.campusLongitude,
        position.latitude,
        position.longitude,
      );

      return distance <= AppConstants.campusRadiusMeters;
    } catch (e) {
      return false;
    }
  }

  // 두 지점 간 거리 계산 (미터)
  double calculateDistance(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    return Geolocator.distanceBetween(lat1, lon1, lat2, lon2);
  }
}

